minor updates for the archiving policy as discussed today in the meeting (#27279)
ocaml/opam-repository: governance/policies/archiving.md

### Compiler cutoff threshold

> The current (2025-02-01) compiler cutoff threshold is `4.08`.

Reviewer: I'm not sure how this date helps. AFAIU, whatever the current threshold is should be stated here, and when it is updates, this should be updated. If someone wants to see when this was written, they can look at git blame.

Reviewer: My worry is that it will lead people to think that this has become "out of date" since this date will probably only change very infrequently. E.g., in 2027, this may still be the current cutoff threshold, but someone is likely to read it and think, "huh, current as of 2025, huh? Well that is quite old, this must be out of date." In other words, I don't think this date answers a questions anyone will have ("current just means current") but I think it is likely to raise irrelevant questions.

That said, it's not a big deal, so if you are sure this is an improvement, I am not really  worried about. If we find it to cause confusion later, we can deal with it at that time.

Reviewer 2: I agree with @hannesm mostly because I am lazy and would never do the git blame 😅 

We have to update the threshold every now and then anyway, at that point we can update also the date. And if we see that it gets stale, we can always drop it later

    - Meaning:
        - Expresses the declared intent of the maintainers to maintain only certain versions ranges of a package.
        - The value of the `x-maintenance-intent` on the latest published package will precedence.

>     - Default value: ["(any)"] on 2025-01-17, please note that this default may change

Reviewer 2: ```suggestion
    - Default value: ["(any)"] on 2025-01-17, please note that this default is expected to change in the future
```
I would make it even more explicit

    - Examples:

>         - `["(latest)"]` the maintainer will only maintain the latest version

Reviewer 2: ```suggestion
        - `["(latest)"]` the maintainer will only maintain the latest version[^caveat]
        [^caveat]: note that this will retain the latest versions of this package so that every supported OCaml version will have an installation candidate.
```
I think we should make this semantic explicit

    - Meaning:
        - Expresses the declared intent of the maintainers to maintain only certain versions ranges of a package.
        - The value of the `x-maintenance-intent` on the latest published package will precedence.
    - Default value: ["(any)"] on 2025-01-17, please note that this default is expected to change in the future
    - Examples:
        - `["(latest)"]` the maintainer will only maintain the latest version[^caveat]

>         [^caveat]: note that this will retain the latest versions of this package so that every supported OCaml version will have an installation candidate.

Reviewer: This caveat is not very clear to me. I am also worried that introducing footnotes make the document harder to read. IMO, we should put all the needed details in the policy directly, without introducing marginal qualifications.

Regarding the content of the caveat: AFAIU the policy, the maintenance intent does not dictate what is "retained". This is determined based on satisfying the inclusion criteria. The intent is way to tell us whether we can just remove a package that fails to satisfy the criteria, or whether we try to contact a maintainer. My understanding of the policy as written is: If a package is no repaired to satisfy the inclusion criteria and the maintainer does not respond within the stated window, the package should be archived no matter what the stated intent was. Conversely, if a package version continues to satisfy the inclusion criteria, it is retained even if it is out-side of the maintenance intent. Given that understanding, I do not understand what this note qualifying retainment relative to ocaml versions adds.

Reviewer: I see that this is a suggestion from @mseri, so tagging.

Author: I struggle what you mean. The whole purpose of archiving things is to reduce the load on opam clients.

Now, if there's package a in version 1.2.3 and 1.2.4 and 1.2.5 with the dependencies all the same, of course we would like to archive 1.2.3 and 1.2.4.

So, the whole addition here is to clarify that a maintenance-intent of latest means that for each supported OCaml compiler, the latest version of that package will be retained. And not only the latest and greatest (in case it requires a newer than the cut-off OCaml compiler). Hope that makes it clearer.

Reviewer: Yes, I am reconsidering too much of the policy here. I retract this comment. We can talk it out at a meeting maybe if it's still worth discussing.

        - `["(none)"]` the maintainer will not maintain any version
        - `["1.3"]` the maintainer will maintain the latest  version of "1.3.Z"
        - `["2.(latest)"]` the maintainer will maintain the latest minor version specifically of version "2" of the package
- `x-maintained`:
    - Allowed values: `true` and `false`
    - Meaning:
        - Expresses that this opam package version is maintained (if `true`) or not (if `false`).
        - Overwrites the `x-maintenance-intent` field
        - Useful to abandon pre-releases

>         - Also for packages that are used in private opam-repositories or developments

Reviewer 2: ```suggestion
        - Also for packages that are used in private opam-repositories or developments and that a maintainer would like to keep in the repository independently of the global package intent
```
I think this is not yet clear enough, maybe this edit? What do you think?

        - `["(none)"]` the maintainer will not maintain any version
        - `["1.3"]` the maintainer will maintain the latest  version of "1.3.Z"
        - `["2.(latest)"]` the maintainer will maintain the latest minor version specifically of version "2" of the package
- `x-maintained`:
    - Allowed values: `true` and `false`
    - Meaning:
        - Expresses that this opam package version is maintained (if `true`) or not (if `false`).

>         - Overwrites the `x-maintenance-intent` field

Reviewer: ```suggestion
        - Overrides the `x-maintenance-intent` field
```

Author: done in df67251905, thanks

>         - Useful to abandon pre-releases

Reviewer: It is not clear to me how this works or what this means. Since the policy document should stand alone, I think expanding this to explain is a good idea.

Author: I clarified in df67251905, hope that is understandable.

>         - Also for packages that are used in private opam-repositories or developments and that a maintainer would like to keep in the repository independently of the global package intent

Reviewer: I am unclear here too. I would not think the opam repository policies would have any bearing on private repositories.

Author: I amended the text in df67251905, is this understandable for you now?

Reviewer: Yes this helps me a lot. Thanks!

    - Examples:
        - `true # used by @bactrian, added on 2025-01-24` where @bactrian wants to keep this package in the opam-repository

>         - `false` if this meets the `x-maintenance-intent`, but is actually not maintained

Reviewer: This seems confusing to me. Why no just update the intent to be accurate, rather than introducing a new field that can conflict with the existing one and then having it override?

Author: I'm sure everyone has opinions, that's why we are meeting to discuss the consensus. Sorry that you couldn't make it last week.

The short story is: there's no "conflict". As example, take the OCaml compiler releases. I don't think anyone will write down an x-maintenance-intent for them, but nevertheless the alpha/beta/rc of old versions can safely be archived.

Reviewer: I sorry I missed the meeting! I'm glad we also have a review process in addition to sync discussions tho :)

More generally, I am worried about the apparent growing complexity of this maintenance data. The more we add new fields and elaborate the bespoke syntax, the more worried I am that we are introducing something that is too esoteric to be that helpful. E.g., if we think we need this `x-maintained` field, maybe we don't need `x-maintenance-intent` at all? Wouldn't it simplify things a lot to just have a boolean flag on each version stating whether or not it is maintained? Then we don't have to worry about inventing a new specification format for all kinds of versions formats :)

Reviewer 2: X-maintenance-intent is global, affects all releases of a package. This allows to override o specify the policy on a specific version of a release. It resolves (manually) all the corner cases we had do far and gives us an alternative to the deprecated flag, which is also package-wide

Reviewer: I believe I understand the intended semantics. But since we've decided a boolean flag is necessary, and now considering how that interacts with, and compares against, the prior `x-maintenance-intent`, I have become very uncertain whether the latter is actually a good solution for the problem we are trying to solve.

I know that this is a bit late in the process to raise this concern, but I think it is better to fix things a little late than it is to stay committed to something that adds complexity and may cause more problems down the line. Here is my current thinking:

Consider: with `X-maintenance-intent`, to determine if any given package version is actually ("intended") to be maintained, you have to:

- find the latest version of that package
- parse out it's `X-maintenance-intent` field into a (from a not fully determined syntax)
- then run the (not fully defined algorithm) to determine if the package version of interest happens to meet the criteria

This requires various forms of knew knowledge and functionality to work correctly, including:

- package publishers to learn 2 new opam fields
- package publishers to learn a novel (albeit simple) DSL for package versions
- us to fully specify a syntax and semantics for this DSL
- the CI and archiving tooling to do back and forth comparing different versions to reason out what is supported.

By contrast, suppose we just have `x-maintained`. To determine if a given package version is maintained, you just:

- check the boolean value of `x-maintained` on the package version

All that is needed to support this is:

- package authors learn one new opam field

So the latter seems like its just a much simpler and more elegant solution to our need to identify whether a package version is maintained. What are the down sides?

The only downside I can think of is that for some packages with a lot of versions, authors may need to make quite a few small updates. But the semantics and logistics of such updates are trivial, and we can very easily offer tooling (e.v., via `opam admin`) to make this easier.

To sum up, thru this discussion, and the surrounding discussions abound the `x-maintenance-intent`, I've come to think that the latter causes more problems that it solves, and that this boolean flag `x-maintained` is all we really need for our purposes. 

I think a principle we can appeal to is that meta-data formats should be as simple as possible, and the corollary that it is better to have simple, unambiguous data formats (with the business logic in tooling to generate and manipulate the data) than it is to enrich the data format with various ad hoc domain-specific logic. I suspect it is much easier for us to write and maintain a tool that can insert a boolean flag into a bunch of opam files, than it will be for our the whole opam-repository socio-technical system to figure out how to manage `x-maintenance-intent`.

WDYT?

Reviewer 2: I am still not convinced. We aim at having a x-maintained-intent with a default that works for most packages and allows us to do a cleanup. I see this is a path of least resistance and least effort.
A tag that needs to be edited continuously would be a massive pain to maintain for us and the maintainers imo. Instead the x-maintained is there for the few corner cases only, and should almost never be needed.

But I think it is worth exploring this space better. Let's have another discussion at the meeting tomorrow

Reviewer: We talked about this at the maintenance meeting today. I am convinced that the current solution is expedient, and the best on the table given the current social practices and the stat of the database and packing system.

If we had a more advanced package management UI, different packaging practices, or support for a different data layout in our existing tooling, then it would make sense for us to pursue a cleaner solution. But given the current state of all this, I am convinced that there will be more value to the ecosystem in adopting a sub-optimal solution that enables us to move forward with the effort than to block trying to achieve a more prefect implementation that would require non-trivial adaptations to our publication process and package tooling.

We also agreed to write a followup PR to clarify some of these points in the policy.

Thank you for the discussion and for your work, @hannesm and thanks for the discussion @mseri.
